aptible/resource_log_drain.go

```diff
@@ -0,0 +1,241 @@
+package aptible
+
+import (
+	"log"
+	"strconv"
+
+	"github.com/aptible/go-deploy/aptible"
+	strfmt "github.com/go-openapi/strfmt"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func resourceLogDrain() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceLogDrainCreate,
+		Read:   resourceLogDrainRead,
+		Delete: resourceLogDrainDelete,
+		Importer: &schema.ResourceImporter{
+			State: resourceLogDrainImport,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"log_drain_id": {
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+			"env_id": {
+				Type:     schema.TypeInt,
+				Required: true,
+				ForceNew: true,
+			},
+			"handle": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"drain_type": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"database_id": {
+				Type:     schema.TypeInt,
+				Optional: true,
+				ForceNew: true,
+			},
+			"drain_username": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+			},
+			"drain_host": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"drain_password": {
+				Type:      schema.TypeString,
+				Optional:  true,
+				Computed:  true, // The API generates a password if one isn't provided
+				ForceNew:  true,
+				Sensitive: true,
+			},
+			"drain_port": {
+				Type:     schema.TypeInt,
+				Optional: true,
+				ForceNew: true,
+			},
+			"logging_token": {
+				Type:      schema.TypeString,
+				Optional:  true,
+				Computed:  true,
+				ForceNew:  true,
+				Sensitive: true,
+			},
+			"url": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"drain_apps": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  true,
+				ForceNew: true,
+			},
+			"drain_databases": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  true,
+				ForceNew: true,
+			},
+			"drain_ephemeral_sessions": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  true,
+				ForceNew: true,
+			},
+			"drain_proxies": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+				ForceNew: true,
+			},
+			// aliases
+			"token": {
+				Type:      schema.TypeString,
```

Comment:
i think if these are specific to type, we can force a validationfunc on these to throw an error if misconfigured (for the wrong type):

ex, using `token` on something that isn't logdna or datadog (- `token` - (alias for `drain_username`) API token used for logdna or datadog), so we enforce the drain_type matching this if possible

Reply:
We should use a `CustomDiff` to perform checks like this a la: https://github.com/hashicorp/terraform-provider-aws/blob/74ee3ac9f3325c117f03da9d568ff4001631a943/internal/service/dynamodb/validate.go#L24.

It would be best to require that the appropriate alias be used as this will provide the best experience since the aliases aren't computed.

Reply:
Actually, since the aliases aren't computed and they're re-hydrated from the API using the field that is aliased, e.g. `drain_username`, would cause the resource to be replaced each apply since the provided alias is `""` but the value it gets back from the API would correspond to the field that is aliased.